elyra/validation/validate.py

```diff
@@ -0,0 +1,424 @@
+        """
+        Validates the pipeline (only pipeline scope) structure for required fields and types
+        :param pipeline: the pipeline definition to be validated
+        :param response: ValidationResponse containing the issue list to be updated
+        """
+
+        current_pipeline_schema_version = '3.0'
+        pipeline_json = json.loads(json.dumps(pipeline))
+
+        # Check for required values.  We require a primary_pipeline, a set of pipelines, and
+        # nodes within the primary pipeline (checked below).
+        if 'primary_pipeline' not in pipeline_json:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Could not determine the primary pipeline.")
+        if not isinstance(pipeline_json["primary_pipeline"], str):
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Field 'primary_pipeline' should be a string.")
+        if 'pipelines' not in pipeline_json:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Pipeline definition not found.")
+        if not isinstance(pipeline_json["pipelines"], list):
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Field 'pipelines' should be a list.")
+
+        primary_pipeline_id = pipeline_json['primary_pipeline']
+        primary_pipeline = PipelineParser._get_pipeline_definition(pipeline_json, primary_pipeline_id)
+        if not primary_pipeline:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Primary pipeline '{}' not found.".format(primary_pipeline_id))
+
+        if 'nodes' not in primary_pipeline or len(primary_pipeline['nodes']) == 0:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="At least one node must exist in the primary pipeline.")
+        if 'app_data' not in primary_pipeline:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Pipeline 'app_data' is missing from primary pipeline")
+        elif 'version' not in primary_pipeline['app_data']:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Pipeline app_data 'version' is missing from primary pipeline")
+        elif not int(primary_pipeline['app_data']['version']) > 0:
+            response.add_message(severity='1', message_type="invalidPipeline",
+                                 message="Primary pipeline version field has an invalid value")
+
+        # TODO: Do we increment the pipeline schema version at all?
+        if pipeline_json['version'] < current_pipeline_schema_version:
```

Comment:
Should the presence of 'version' be confirmed prior to this - like the others are doing?

Reply:
updated